fdbclient/include/fdbclient/TenantEntryCache.actor.h

```diff
@@ -102,16 +108,158 @@ class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyab
 		return tenantList.results;
 	}
 
+	ACTOR static Future<Void> refreshCacheById(int64_t tenantId,
+	                                           TenantEntryCache<T>* cache,
+	                                           TenantEntryCacheRefreshReason reason) {
+		TraceEvent(SevInfo, "TenantEntryCacheIDRefreshStart", cache->id()).detail("Reason", static_cast<int>(reason));
+		state Reference<ReadYourWritesTransaction> tr = cache->getDatabase()->createTransaction();
+		loop {
+			try {
+				tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::READ_LOCK_AWARE);
+				state Optional<TenantName> name = wait(TenantMetadata::tenantIdIndex().get(tr, tenantId));
+				if (name.present()) {
+					Optional<TenantMapEntry> entry = wait(TenantMetadata::tenantMap().get(tr, name.get()));
+					if (entry.present()) {
+						cache->put(std::make_pair(name.get(), entry.get()));
+						updateCacheRefreshMetrics(cache, reason);
+					}
+				}
+				break;
+			} catch (Error& e) {
+				wait(tr->onError(e));
+			}
+		}
+		TraceEvent(SevInfo, "TenantEntryCacheIDRefreshEnd", cache->id()).detail("Reason", static_cast<int>(reason));
+		return Void();
+	}
+
+	ACTOR static Future<Void> refreshCacheByName(TenantName name,
+	                                             TenantEntryCache<T>* cache,
+	                                             TenantEntryCacheRefreshReason reason) {
+		TraceEvent(SevInfo, "TenantEntryCacheNameRefreshStart", cache->id()).detail("Reason", static_cast<int>(reason));
+		state Reference<ReadYourWritesTransaction> tr = cache->getDatabase()->createTransaction();
+		loop {
+			try {
+				tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::READ_LOCK_AWARE);
+				Optional<TenantMapEntry> entry = wait(TenantMetadata::tenantMap().get(tr, name));
+				if (entry.present()) {
+					cache->put(std::make_pair(name, entry.get()));
+					updateCacheRefreshMetrics(cache, reason);
+				}
+				break;
+			} catch (Error& e) {
+				wait(tr->onError(e));
```

Comment:
Is there an opportunity to avoid code-duplication between `refreshCacheById` and `refreshCacheByName`?

Reply:
I thought about it, it's a bit tricky since they use different system keys so everything after setting the transaction options is different. The only think we could really refactor out is the setting up of the transaction which im not sure if it's worth it